sklearn/preprocessing/_discretization.py

```diff
@@ -185,7 +185,10 @@ def fit(self, X, y=None):
 
             # Remove redundant bins (i.e., bins whose width = 0)
             if self.strategy in ('quantile', 'kmeans'):
-                bin_edges[jj] = np.unique(bin_edges[jj])
+                bin_edges[jj] = np.array(
+                    [bin_edges[jj][0]] +
+                    [bin_edges[jj][i] for i in range(1, len(bin_edges[jj]))
+                     if bin_edges[jj][i] - bin_edges[jj][i - 1] > 1e-8])
```

Comment:
we would still be solving both issues if we had >0 here. Are we sure we want a hard coded tolerance of 1e-8?

Reply:
Is `mask = np.ediff1d(bin_edges[jj], to_begin=np.inf) > 1e-8; bin_edges[jj] = bin_edges[jj][mask]` clearer?

Reply:
I think we need 1e-8 here, see e.g., 
```
import numpy as np
a = np.ones(279) * 0.56758051638767337
ans = []
for p in range(0, 100, 5):
    ans.append(np.percentile(a, p))
print([ans[0]] + [ans[i] for i in range(1, len(ans)) if ans[i] - ans[i - 1] > 0])
# [0.5675805163876734, 0.5675805163876735]
print([ans[0]] + [ans[i] for i in range(1, len(ans)) if ans[i] - ans[i - 1] > 1e-8])
# [0.5675805163876734]
```
Though we should blame numpy for this

Reply:
> Is `mask = np.ediff1d(bin_edges[jj], to_begin=np.inf) > 1e-8; bin_edges[jj] = bin_edges[jj][mask]` clearer?

At least not from my side :) But I assume this can make use of some optimization in numpy.